Add some forward AD formulas

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69727
* #69740
* #69690
* #69661
* #69597
* __->__ #69384
* #69592
* #69558

Differential Revision: [D33020602](https://our.internmc.facebook.com/intern/diff/D33020602)

soulitzer committed:
Add auto_element_wise formulas

[ghstack-poisoned]

soulitzer committed:
Update on "[IGNORE] Add auto_element_wise formulas"




[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10005,17 +10012,27 @@ def ref_pairwise_distance(input1, input2):
            skips=(
                # Pre-existing condition; Needs to be fixed
                DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_composite_compliance'),
+               # This is a composite op that `sort` enables, but got: Trying to set a forward gradient
```
Is that because the sort formula is wrong?

soulitzer replied:
I think it was actually failing because of the batched forward grad check

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10231,8 +10248,11 @@ def ref_pairwise_distance(input1, input2):
+           supports_forward_ad=True,
            sample_inputs_func=sample_inputs_as_strided,
            skips=(
+               # RuntimeError: The size of tensor a (3) must match the size of tensor b (2) at non-singleton dimension
```
We should have an inplace formula then that does the right thing? (I guess doing inplace as strided instead of copy_)

albanD reviewed (commented):
Sounds mostly good

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -2342,6 +2369,7 @@
 
 - name: unbind.int(Tensor(a -> *) self, int dim=0) -> Tensor(a)[]
   self: unbind_backward(grads, dim)
+  result: auto_linear
```
TODO: need to test this

soulitzer committed:
Update on "Add some forward AD formulas"




[ghstack-poisoned]

soulitzer committed:
Update on "Add some forward AD formulas"




[ghstack-poisoned]

soulitzer committed:
Update on "Add some forward AD formulas"




[ghstack-poisoned]

soulitzer committed:
Update on "Add some forward AD formulas"




[ghstack-poisoned]

albanD reviewed (approved):
Looks good!

soulitzer committed:
Update on "Add some forward AD formulas"




[ghstack-poisoned]

soulitzer committed:
Update on "Add some forward AD formulas"




[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020602).

soulitzer committed:
Update on "Add some forward AD formulas"


Differential Revision: [D33020602](https://our.internmc.facebook.com/intern/diff/D33020602)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020602).

soulitzer committed:
Update on "Add some forward AD formulas"


Differential Revision: [D33020602](https://our.internmc.facebook.com/intern/diff/D33020602)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020602).

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020602).
